### PR TITLE
Move REGISTERED_DIAGNOSTICS to a ParseSess field

### PR DESCRIPTION
**Pull request by Zoxc**
r? @michaelwoerister

### CODE REVIEW

File: src/libsyntax/diagnostics/plugin.rs [hunk 1]
```diff
@@ -132,17 +117,17 @@ pub fn expand_register_diagnostic<'cx>(ecx: &'cx mut ExtCtxt,
...
-            description,
-            use_site: None
-        };
-        if diagnostics.insert(code.name, info).is_some() {
-            ecx.span_err(span, &format!(
-                "diagnostic code {} already registered", code
-            ));
-        }
-    });
+    let mut diagnostics = ecx.parse_sess.registered_diagnostics.lock();
```
Comment (michaelwoerister) [inline src/libsyntax/diagnostics/plugin.rs]: Could we add a `with` method to `Lock`? That would make it clearer in many cases whether the lock is held until the end of the block intentionally or not.

**Review by michaelwoerister — COMMENTED** [review]
Looks good to me. However, I think it would really be an improvement if we had a `Lock::with()` method as that would encourage making locking ranges more visible and explicit.

File: src/libsyntax/diagnostics/plugin.rs (hunk 1 above)
Comment (Zoxc) [reply src/libsyntax/diagnostics/plugin.rs]: That's probably not a bad idea. `with` sounds a little innocent for a lock though.

Comment (michaelwoerister) [reply src/libsyntax/diagnostics/plugin.rs]: `with_lock()` maybe?

**Comment by Zoxc** [comment]
I added methods to lock in a closure.

**Comment by michaelwoerister** [comment]
:+1: 

@bors r+

**Pull request merged**